lib/Runtime/Base/CompactCounters.h

```diff
@@ -235,45 +164,42 @@ namespace Js
             {
                 if (sizeof(FieldT) == 2)
                 {
-                    for (; i < signedStart; i++)
-                    {
-                        fieldsArray->u16Fields[i] = oldFieldsArray->u8Fields[i];
-                    }
                     for (; i < max; i++)
                     {
-                        fieldsArray->i16Fields[i] = oldFieldsArray->i8Fields[i];
+                        fieldsArray->u16Fields[i] = oldFieldsArray->u8Fields[i];
                     }
                 }
                 else if (sizeof(FieldT) == 4)
                 {
-                    for (; i < signedStart; i++)
-                    {
-                        fieldsArray->u32Fields[i] = oldFieldsArray->u8Fields[i];
-                    }
                     for (; i < max; i++)
                     {
-                        fieldsArray->i32Fields[i] = oldFieldsArray->i8Fields[i];
+                        fieldsArray->u32Fields[i] = oldFieldsArray->u8Fields[i];
                     }
                 }
             }
             else if (this->fieldSize == 2)
             {
-                for (; i < signedStart; i++)
-                {
-                    fieldsArray->u32Fields[i] = oldFieldsArray->u16Fields[i];
-                }
                 for (; i < max; i++)
                 {
-                    fieldsArray->i32Fields[i] = oldFieldsArray->i16Fields[i];
+                    fieldsArray->u32Fields[i] = oldFieldsArray->u16Fields[i];
                 }
             }
             else
             {
                 Assert(this->fieldSize==0);
             }
 
-            this->fieldSize = sizeof(FieldT);
-            this->fields = fieldsArray;
+            if (this->fieldSize == 0)
+            {
+                this->fieldSize = sizeof(FieldT);
+                this->fields = fieldsArray;
+            }
+            else
+            {
+                AutoCriticalSection autoCS(ThreadContext::GetThreadContextList()->GetEtwRundownCriticalSection());
```

Comment:
Careful here- you already hold the ThreadContext's critical section- please trace through the code and make sure you don't end up deadlocking